File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/ast/MultilineDocStyleRule.java
package com.ririnto.sinon.harness.rules.ast;

import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.comments.JavadocComment;
import com.ririnto.sinon.harness.Finding;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import org.apache.maven.plugin.MojoExecutionException;
import tools.jackson.databind.JsonNode;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;

/**
 * Rule that requires JavaDoc comments to use multiline block style.
 */
public enum MultilineDocStyleRule implements AstRule {
    INSTANCE;

    private static final String CATEGORY = "multilineDocStyle";

    @Override
    public String category() {
        return CATEGORY;
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        if (!"multiline".equals(docStyleMode(ctx.manifest().raw()))) {
            return List.of();
        }
        final String severity = HarnessCheckHelper.getSeverity(ctx.manifest().raw(), CATEGORY);
        return ctx.stackSources(CATEGORY).stream()
                .flatMap(file -> validateFile(ctx.root(), file, severity).stream())
                .toList();
    }

    private List<Finding> validateFile(Path root, Path file, String severity) {
        try {
            final CompilationUnit compilationUnit = StaticJavaParser.parse(file);
            return compilationUnit.getAllContainedComments().stream()
                    .filter(comment -> comment instanceof JavadocComment)
                    .filter(comment -> !comment.toString().contains("\n"))
                    .map(comment -> Finding.of(severity, CATEGORY, root.relativize(file) + ":" + comment.getBegin().map(position -> position.line).orElse(1) + ": JavaDoc comment must use multiline style"))
                    .toList();
        } catch (IOException error) {
            return List.of(Finding.of(severity, CATEGORY, "failed to parse " + root.relativize(file) + ": " + error.getMessage()));
        }
    }

    private static String docStyleMode(JsonNode manifest) {
        final JsonNode section = manifest.get(CATEGORY);
        if (section == null || !section.has("parameters")) {
            return "multiline";
        }
        final JsonNode mode = section.get("parameters").get("docStyleMode");
        return mode == null ? "multiline" : mode.asText("multiline");
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/ast/MultilineDocStyleRule.kt
package com.ririnto.sinon.harness.rules.ast

import com.ririnto.sinon.harness.ast.AstFinding
import com.ririnto.sinon.harness.ast.AstFindingRenderer
import com.ririnto.sinon.harness.ast.AstSupport
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessAstRule
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.jetbrains.kotlin.psi.KtDeclaration
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.KtPsiFactory
import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
import java.nio.file.Path

/**
 * Rule that requires KDoc comments to use multiline block style.
 */
object MultilineDocStyleRule : HarnessAstRule() {
    /**
     * Category key.
     */
    override val category: String = "multilineDocStyle"

    override fun renderAstFindings(
        ctx: RuleContext,
        findings: Collection<AstFinding>,
    ): Collection<Finding> = AstFindingRenderer.renderEach(findings.toList(), ctx.manifest.raw)

    override fun findAstFindings(
        file: Path,
        ctx: RuleContext,
        astFactory: KtPsiFactory?,
    ): Collection<AstFinding> =
        buildSet {
            if (docStyleMode(ctx) != "multiline") {
                return@buildSet
            }
            val ktFile = AstSupport.parse(file, astFactory)
            ktFile?.accept(Visitor(::add, file, ctx, ktFile))
        }

    private fun docStyleMode(ctx: RuleContext): String =
        ctx.manifest
            .categoryObject(category)
            ?.get("parameters")
            ?.jsonObject
            ?.get("docStyleMode")
            ?.jsonPrimitive
            ?.contentOrNull
            ?: "multiline"

    private class Visitor(
        private val record: (AstFinding) -> Unit,
        private val file: Path,
        private val ctx: RuleContext,
        private val ktFile: KtFile,
    ) : KtTreeVisitorVoid() {
        override fun visitDeclaration(declaration: KtDeclaration) {
            super.visitDeclaration(declaration)
            val docComment = declaration.docComment ?: return
            if (!docComment.text.contains('\n')) {
                record(
                    AstFinding(
                        rule = "multilineDocStyle",
                        file = AstSupport.relativeFilePath(file, ctx.root),
                        line = AstSupport.lineOf(ktFile, docComment.node?.startOffset),
                        details = mapOf("style" to "kdoc"),
                    ),
                )
            }
        }
    }
}
